### DISCUSSION TITLE
aws cli error handling

### DISCUSSION THREAD

**Discussion by sourdoughs** [community/community#63679] [Actions] [answered]
### Select Topic Area

Question

### Body

I have a script which retrieves parameter keys in aws ssm. I would like this loop to continue even when a parameter is not found as I have an array of parameters and sometimes not all of it will be needed for an application build. I am unable to handle this scenario after trying multiple approaches. What could be wrong here?

here's a snippet of the script:

PARAMETER_NAMES=("somekey" "somekey1" "somekey2")
SECRET_KEYS("somevalue" "somevalue1" "somevalue2")

for i in "${!PARAMETER_NAMES[@]}"; do
    param="${PARAMETER_NAMES[i]}"
    secret_key="${SECRET_KEYS[i]}"

# Attempt to get the parameter value
    value=$(aws ssm get-parameter --name "$param" --query "Parameter.Value" --output text 2>&1)

    # Check the error message for "ParameterNotFound" and continue if not found
    if [[ "$value" == *"ParameterNotFound"* ]]; then
        echo "Parameter $param not found, skipping..."
    else
        echo "$param=$value" >> parameters.env
        echo "::add-mask::$value"  # Mask the value in workflow logs
    fi
done

This doesn't work, it still exists with 254.

**Comment by nirav-gajera** [comment]
The issue with your script is that you are using the `if` statement to check for the error message "ParameterNotFound". However, this error message is only returned if the parameter does not exist at all. If the parameter does exist, but the value is empty, the `aws ssm get-parameter` command will return an empty string.

To fix this, you can use the `while` loop to iterate over the array of parameters. In the `while` loop, you can use the `test` command to check if the parameter value is empty. If the parameter value is empty, you can skip the parameter and continue to the next one.

Here is a modified version of your script that fixes the issue:

```
PARAMETER_NAMES=("somekey" "somekey1" "somekey2")
SECRET_KEYS=("somevalue" "somevalue1" "somevalue2")

i=0
while [ "$i" -lt "${#PARAMETER_NAMES[@]}" ]; do
  param="${PARAMETER_NAMES[$i]}"
  secret_key="${SECRET_KEYS[$i]}"

  value=$(aws ssm get-parameter --name "$param" --query "Parameter.Value" --output text 2>&1)

  # Check if the parameter value is empty
  if test -z "$value"; then
    echo "Parameter $param is empty, skipping..."
  else
    echo "$param=$value" >> parameters.env
    echo "::add-mask::$value"  # Mask the value in workflow logs
  fi

  i=$(($i + 1))
done
```

**Reply by sourdoughs** [reply]
Hi, Thanks for you reply. But the ParameterNotFound is the only scenario that I want to skip. All others like if it's an empty string, the loop should fail.

**Reply by nirav-gajera** [reply]
oh for that, you can use the `or` operator to check for both the "ParameterNotFound" error message and an empty string. The `or` operator will return true if either condition is true.

ex:

```
PARAMETER_NAMES=("somekey" "somekey1" "somekey2")
SECRET_KEYS=("somevalue" "somevalue1" "somevalue2")

i=0
while [ "$i" -lt "${#PARAMETER_NAMES[@]}" ]; do
  param="${PARAMETER_NAMES[$i]}"
  secret_key="${SECRET_KEYS[$i]}"

  value=$(aws ssm get-parameter --name "$param" --query "Parameter.Value" --output text 2>&1)

  # Check if the parameter value is empty or if the error message is ParameterNotFound
  if [[ "$value" == *"ParameterNotFound"* ]] || [ -z "$value" ]; then
    echo "Parameter $param is empty or not found, skipping..."
  else
    echo "$param=$value" >> parameters.env
    echo "::add-mask::$value"  # Mask the value in workflow logs
  fi

  i=$(($i + 1))
done
```

hope it works

**Reply by sourdoughs** [reply]
This still did not work. For some reason the script is unable to handle that response from aws (An error occurred (ParameterNotFound) when calling GetParameter operation:) when the parameter is not present.

**Reply by nirav-gajera** [reply]
Try upgrading your AWS cli client. maybe There's a bug that was floating around and your version might be affected.

**Reply by sourdoughs** [reply]
I'm not sure it's the aws cli that is the issue. This actually works on my local and tested with 2.7 and 2.13 versions of the cli. both works. 
![Screenshot 2023-08-14 at 8 12 06 PM](https://github.com/community/community/assets/46431314/d52533ed-b022-4fb9-b19b-59f24476d0d8)

I'm using a self hosted runner though with ubuntu.

**Comment by sourdoughs** [comment, marked as answer]
I got this to work by moving the script to a .sh file and calling it from within the yml.
